### PR TITLE
Exercise 2

### PR DESCRIPTION
**Pull request by MaykAkifovski**

### CODE REVIEW

File: webapp/.storybook/preview.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const parameters = {
+  actions: { argTypesRegex: "^on[A-Z].*" },
+}
```
Comment (aloparev) [inline webapp/.storybook/preview.js]: As mentioned in the lecture, this can be removed.
```suggestion
export const parameters = {

}
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: CI
+
+on:
+  pull_request:
+    branches: [ main ]
```
Comment (roschaefer) [inline .github/workflows/main.yml]: Why only branch `main`? The homework task says:
>   - update [status checks](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-status-checks) on all PRs to your repository

File: webapp/src/components/news-form/NewsForm.stories.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import Vue from 'vue';
+import Vuetify from 'vuetify';
+import NewsForm from "./NewsForm.vue";
+
+Vue.use(Vuetify);
+
+/*
+TODO:
+fix "this.uniqueTitles" is undefined error -> NewsForm rules
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.stories.js:9]: You can fix it like this:
```js
const Template = (args, {argTypes}) => ({
    components: { NewsForm },
    template: '<news-form v-bind="$props"/>',
    props: Object.keys(argTypes)
});

export const Default = Template.bind({}); 
Default.args = {
  uniqueTitles: [] // you can create another story with a non empty list as default value
}
```

File: webapp/src/components/news-form/NewsForm.test.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	/*
+	TODO: 
+	- understand why 'valid' does not change value when setData() is executed
+	- afterwards refactor the test
+	*/
+	
+	it('should accept less than 64 character', async () => {
+
+	const wrapper = mount(NewsForm, {
+        data() {
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: You shouldn't need to set `newsTitle` here.

File: webapp/src/components/news-form/NewsForm.test.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            return {
+				newsTitle: ''
+            }
+        }
+    });
+      
+    await wrapper.setData({
+		valid: true,
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: You're mocking too much data here!

File: webapp/src/components/news-form/NewsForm.test.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        newsTitle: itemTitle
+    });
+
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js:36]: What about:
```js
await wrapper.find('input').trigger('input', 'your cool title')
```
Or:
```js
await wrapper.find('input').setValue('your cool title')

```
See: https://vue-test-utils.vuejs.org/api/wrapper/#setvalue

File: webapp/src/components/news-form/NewsForm.test.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		valid: true,
+        newsTitle: itemTitle
+    });
+	
+	//await wrapper.vm.$nextTick();
+      
+	expect(wrapper.vm.newsTitle).toBe(itemTitle);
+    expect(wrapper.vm.valid).toBe(true);
+
+    wrapper.destroy();
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: No need for that, instead re-initialize `wrapper` before every test case:
```js
beforeEach(() => {
  wrapper = mount(/*...*/)
})

File: webapp/src/components/news-form/NewsForm.test.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    });
+	
+	it('should not accept more than 64 character', async () => {
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]:
```suggestion
	it('does not accept more than 64 character', async () => {
```
Be affirmative

File: webapp/src/components/news-form/NewsForm.test.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      
+    await wrapper.setData({
+		valid: true,
+        newsTitle: itemTitle
+    });
+	
+	//await wrapper.vm.$nextTick();
+      
+	expect(wrapper.vm.newsTitle).toBe(itemTitle);
+    expect(wrapper.vm.valid).toBe(true);
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: Split these two assertions into two test cases.

File: webapp/src/components/news-form/NewsForm.test.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    });
+      
+    await wrapper.setData({
+		valid: true,
+        newsTitle: itemTitle
+    });
+	
+	//await wrapper.vm.$nextTick();
+      
+	expect(wrapper.vm.newsTitle).toBe(itemTitle);
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: Don't test internal properties. Instead check what you can see:
```js
expect(wrapper.find('input').attributes('value')).toEqual('Fake News')
```
Also use hard-coded assertion values to avoid silently passing false-negative test cases.

File: webapp/src/components/news-form/NewsForm.test.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    expect(wrapper.vm.valid).toBe(true);
+
+    wrapper.destroy();
+    });
+	
+	it('should not accept more than 64 character', async () => {
+
+	const wrapper = mount(NewsForm, {
+        data() {
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.test.js]: As I said: The reactive state `data` is not part of the public interface of a component. You shouldn't need to set it in your tests.

File: webapp/src/components/news-form/NewsForm.vue [hunk 12]
```diff
@@ -4,6 +4,7 @@
       <v-row>
         <v-col>
           <v-text-field
+			id="newsTitle"
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.vue]: Don't use HTML ids as they should be unique on a website. If you render the component twice, you would violate that agreement. You could use attributes or classes.

File: webapp/src/components/news-form/NewsForm.vue [hunk 13]
```diff
@@ -15,6 +16,7 @@
       <v-row>
         <v-col>
           <v-btn
+			id="adder"
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.vue]: If you want, you can use an attribute `data-test-*`. Some people even remove those attributes when they build for production: https://forum.vuejs.org/t/how-to-remove-attributes-from-tags-inside-vue-components/24138/3?u=linusborg

File: webapp/src/components/news-form/NewsForm.vue [hunk 14]
```diff
@@ -31,7 +33,7 @@
 <script>
 export default {
   name: 'NewsForm',
-  'props': ['uniqueTitles'],
+  props: ['uniqueTitles'],
```
Comment (roschaefer) [inline webapp/src/components/news-form/NewsForm.vue]: It's best practice to do:
```suggestion
  props: {
    uniqueTitles: { type: Array, required: true }
  }
```
So you always get a nice warning when you render your component and forgot to set the properties (which you did in your storybook story).

File: webapp/src/components/news-item/NewsItem.stories.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import Vue from 'vue';
+import Vuetify from 'vuetify';
+import NewsItem from "./NewsItem.vue";
+
+Vue.use(Vuetify);
+
+export default {
+    title: 'Components/NewsItem',
+    component: NewsItem,  
+    argTypes: {
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.stories.js:10]:
```suggestion
    argTypes: {
      upvote: { action: 'upvote' },
```

And you can do 
```
    template: '<news-item :newsItem="newsItem" @upvote="upvote" />',
```
on line 18 to set up event handlers for your events in your storybook.

File: webapp/src/components/news-item/NewsItem.test.js [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+          title: title,
+          votes: votes,
+        },
+      },
+    });
+
+    expect(wrapper.props().newsItem.title).toBe(title);
+    expect(wrapper.props().newsItem.votes).toBe(votes);
+
+  });
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.test.js]:
```suggestion
```
This is the already tested behaviour of vue-test-utils. No need to test that again.

File: webapp/src/components/news-item/NewsItem.test.js [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        },
+      },
+    });
+
+    const expectedText = title + ' ' + votes;
+
+    const actualH1 = wrapper.find('h1');
+    const actualUpvoteButton = wrapper.find('button.upvote');
+    const actualDownvoteButton = wrapper.find('button.downvote');
+    const actualRemoveButton = wrapper.find('button.remove');
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.test.js:31]: You could probably make a separate test case for each of these.

File: webapp/src/components/news-item/NewsItem.test.js [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+    expect(actualH1).toBeDefined();
+    expect(actualH1.text()).toEqual(expectedText);
+
+    expect(actualUpvoteButton).toBeDefined();
+    expect(actualDownvoteButton).toBeDefined();
+    expect(actualRemoveButton).toBeDefined();
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.test.js]: Use: https://vue-test-utils.vuejs.org/api/wrapper/#exists

File: webapp/src/components/news-item/NewsItem.test.js [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+          title: TITLE,
+          votes: VOTES,
+        },
+      },
+    });
+
+    const button = wrapper.find('button.downvote');
+    button.trigger('click');
+
+    await wrapper.vm.$nextTick();
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.test.js]:
```suggestion
    await button.trigger('click');
```

File: webapp/src/components/news-item/NewsItem.test.js [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+    const expectedCallCount = 1;
```
Comment (roschaefer) [inline webapp/src/components/news-item/NewsItem.test.js:81]:
```suggestion
```
remove and use a hard-coded value on line 103

File: webapp/src/components/news-list/NewsList.test.js [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+      wrapper.destroy();
+  })
+
+  it("should switch the sorting order whenever the button is clicked", () => {
+      const wrapper = mount(NewsList, {
+          data() {
+              return {
+                  newsItems: [
+                      {title: "Title_1", votes: 5},
+                      {title: "Title_2", votes: 4}
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.test.js]: If you need to change some options of your test setup but you want to re-use the setup:

```js
let = setupWrapper(options) => {
  const defaults = { propsData: { newsItems: [] } }
  return mount(NewsList, { ...defaults, ...options })
}

// .. later in your tests:
const wrapper = setupWrapper({propsData: { newsItems: [{title: "Title_1", votes: 5}]} })

File: webapp/src/components/news-list/NewsList.test.js [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+                  descending: true,
+              }
+          }
+      });
+
+      const item = wrapper.findComponent(NewsItem);
+
+      item.find('button.upvote').trigger('click');
+
+      await item.vm.$nextTick();
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.test.js]:
```suggestion
      await item.find('button.upvote').trigger('click');
```

File: webapp/src/components/news-list/NewsList.test.js [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+
+      expect(item.emitted('updateNews')).toBeTruthy();
+      expect(item.emitted('updateNews').length).toBe(1);
+
+      expect(wrapper.vm.newsItems[0].votes).toBe(6);
+
+      wrapper.destroy();
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.test.js]: Avoid a lot of code (and potential bugs) and re-initialize your wrapper in a `beforeEach`!

File: webapp/src/components/news-list/NewsList.test.js [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+      const item = wrapper.findComponent(NewsItem);
+
+      item.find('button.upvote').trigger('click');
+
+      await item.vm.$nextTick();
+
+      expect(item.emitted('updateNews')).toBeTruthy();
+      expect(item.emitted('updateNews').length).toBe(1);
+
+      expect(wrapper.vm.newsItems[0].votes).toBe(6);
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.test.js]:
```suggestion
      expect(item.emitted()).toEqual([['updatenews', {title: "Title_1", votes: 6},]])
```

File: webapp/src/components/news-list/NewsList.test.js [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+                  newsItems: [
+                      {title: "Title_1", votes: 5},
+                      {title: "Title_2", votes: 4}
+                  ],
+                  descending: true,
+              }
+          }
+      });
+
+      const item = wrapper.findComponent(NewsItem);
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.test.js]: You should test this feature in `NewsItem.test.js`, not in the parent component.

File: webapp/src/components/news-list/NewsList.vue [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    addItem(title) {
+      this.newsItems.push({ title: title, votes: 0 });
+    },
+    toggleSortOrder() {
+      this.descending = !this.descending;
+    },
+    noNews: function() {
+      return this.newsItems.isEmpty();
+    },
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.vue]:
```suggestion
```
`isEmpty()` is not defined on Array

File: webapp/src/components/news-list/NewsList.vue [hunk 27]
```diff
@@ -0,0 +1,85 @@
+<template>
+  <div>
+    <div v-if="sortedItems.length == 0">
```
Comment (roschaefer) [inline webapp/src/components/news-list/NewsList.vue]:
```suggestion
    <div v-if="sortedItems.length">
```

Comment (roschaefer) [reply webapp/src/components/news-list/NewsList.vue]: it's ugly but in JS `!! 0` is `false`

File: webapp/src/stories/Button.stories.js [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+Large.args = {
+  size: 'large',
+  label: 'Button',
+};
+
+export const Small = Template.bind({});
+Small.args = {
+  size: 'small',
+  label: 'Button',
+};
```
Comment (roschaefer) [inline webapp/src/stories/Button.stories.js]: Remove unrelated changes!

**Review by roschaefer — CHANGES_REQUESTED** [review]
lot of text

1. Please remove unrelated code
2. Please try to be a little more concise and write *less* code :stuck_out_tongue: 

I haven't even checked out your code and it took me a long time to read. Hope my suggestions are of some help.

File: .github/workflows/main.yml (hunk 2 above)
Comment (aloparev) [reply .github/workflows/main.yml]: You can hook this just to push or pull requests, independent from the branch.
```suggestion
    on: [push, pull_request]
```

File: webapp/src/components/news-item/NewsItem.vue [hunk 29]
```diff
@@ -1,9 +1,9 @@
 <template>
   <div>
     <h1>{{ newsItem.title }} {{ (newsItem.votes) }}</h1>
-    <button @click="incrementVotes">Upvote</button>
-    <button @click="decrementVotes">Downvote</button>
-    <button @click="removeMe">Remove</button>
+    <button class="upvote" @click="incrementVotes">Upvote</button>
+    <button class="downvote" @click="decrementVotes">Downvote</button>
+    <button class="remove" @click="removeMe">Remove</button>
```
Comment (aloparev) [inline webapp/src/components/news-item/NewsItem.vue:6]: Using additional classes to later address the buttons is very handy.

**Review by aloparev — CHANGES_REQUESTED** [review]
Looks great for my untrained eye, you even implemented user specific header. Why not use ready to use .css like Bootstrap though?

File: webapp/src/components/news-item/NewsItem.test.js (hunk 20 above)
Comment (MaykAkifovski) [reply webapp/src/components/news-item/NewsItem.test.js:81]: I find it more readable with a variable.

**Review by aloparev — APPROVED** [review]

File: .github/workflows/main.yml [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Install Dependencies
+        run: npm install
+
+      - name: Check lint
+        run: npm run lint
+
+      - name: Run Tests
+        run: npm run test
```
Comment (medizinmensch) [inline .github/workflows/main.yml:26]: :star:   for setting up a build server that tests and lints your code

File: webapp/src/components/news-form/NewsForm.stories.js [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+    title: 'Components/NewsForm', 
+};
+
+const Template = (args, {argTypes}) => ({
+    components: { NewsForm },
+    template: '<news-form v-bind="$props"/>',
+    props: Object.keys(argTypes)
+});
+
+export const Default = Template.bind({});
```
Comment (medizinmensch) [inline webapp/src/components/news-form/NewsForm.stories.js:22]: :star:   for creating a story for all of your components, locating them correcctly and   adding the @storybook/addon-essentials along with @storybook/addon-a11y

File: webapp/src/components/news-form/NewsForm.test.js [hunk 32]
```diff
@@ -0,0 +1,63 @@
+import {describe, expect, it, beforeEach} from "@jest/globals";
+import {mount, createLocalVue} from "@vue/test-utils";
```
Comment (medizinmensch) [inline webapp/src/components/news-form/NewsForm.test.js:2]: :star:   for testing your components with jest and vue-test-utils

File: webapp/src/components/news-list/NewsList.test.js [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+    }) 
+
+    it('should show the message "The list is empty :(" when there are no elements in the list', () => {
+      const wrapper = setupWrapper();
+
+      const paragraph = wrapper.find("h2");
+
+      expect(paragraph.text()).toEqual("The list is empty :(");
+      wrapper.destroy();
+  })
```
Comment (medizinmensch) [inline webapp/src/components/news-list/NewsList.test.js:37]: :star:   for the empty-list-feature and it's test

File: webapp/src/components/news-list/NewsList.test.js [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+
+  it("should switch the sorting order whenever the button is clicked", () => {
+
+    const button = wrapper.find('button.sorter');
+    button.trigger('click');
+
+    let firstDomElementInAscendingMode = wrapper.find('h1');
+    expect(firstDomElementInAscendingMode.text()).toContain("4");
+  })
```
Comment (medizinmensch) [inline webapp/src/components/news-list/NewsList.test.js:51]: :star:   for the toggle between ascending and desccending order and it's test

File: webapp/package.json [hunk 35]
```diff
@@ -46,5 +81,24 @@
...
+      ".*\\.(js)$": "babel-jest"
+    },
+    "moduleNameMapper": {
+      "^@/(.*)$": "<rootDir>/src/$1"
+    },
+    "collectCoverage": true,
+    "collectCoverageFrom": [
+      "**/*.{js,vue}",
+      "!**/node_modules/**"
+    ]
```
Comment (medizinmensch) [inline webapp/package.json:102]: :star:   for configuring jest in a way so you can locate component, story and spec in   one folder

**Review by medizinmensch — COMMENTED** [review]
See :star: as below.

Also:
* :star:  for requesting PR review as well as reviewing another teams code

Sadly, as I checked our your code and tried to run it, I got several issues: 
* the `npm run test` gives my some errors and 50+ warnings. This leads to at least one test failing (see output below). I will have to deduct a :x: for that. 
```
Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 11 passed, 12 total
Snapshots:   0 total
Time:        9.028 s
```
* you installed several storybook addons into your package.json, but you didn't configure them for storybooks. This means you are not using the controls feature as well as the a11y features. Furthermore you installed `@storybook/addon-knobs` which, as it was explicitly stated, shall not be installed. I have to deduct a point for not passing the accessibility tests :x: and one point for setting yp your component properties in realtime with the `@storybook/addon-ccontrols` :X:


That makes 7/10 points! 

good!
